### PR TITLE
refactor(web):  new abstraction - corrective layouts 🐵

### PR DESCRIPTION
**Pull request by jahorton**
With the new incoming types of gestures, it's important that we prepare a reasonable way to model them alongside the predictive-text engine.  While we've long had a good setup in place for standard keyboard layers... none of the other uses cases have particularly great models.

Our existing longpress solution:

https://github.com/keymanapp/keyman/blob/af3ba1f07dd806638ed907a33a16ebf107b721aa/web/src/engine/osk/src/visualKeyboard.ts#L578-L583

Yeah, that's not ideal.  Surely we can do better.

So, to provide a good path forward for our existing longpresses and for newly-incoming multitaps and flicks, this PR creates abstractions for the minimal requirements of the key-distance algorithm at the core of the fat-finger algorithm.  It also extracts the key-correction methods from `ActiveLayer`, allowing them to operate independently of any specific type - they'll now utilize the new abstractions.

By doing so, it becomes possible to build appropriate values for the new abstraction for our subkey menus, allowing _true_ fat-finger calculations for subkeys.  (A major improvement over simply maintaining the existing hack.)  It also provides a baseline for further enhancements we may wish to consider for corrections involving multitaps and flicks.  Note that none of the ideas in this paragraph are implemented within this PR, though - I've kept the scope limited to just the refactor needed to facilitate implementations thereof.

An extra benefit from this refactor - the key-distance types and algorithms at the heart of the fat-finger algorithm have been moved to `input-processor`, the same package containing the fat-finger algorithm.  As `input-processor`'s primary role is facilitating predictive-text and acting as the interface between keystroke-engine and predictive-text, it's the perfect place for them.  (Not that `input-processor` is the perfect _name_, given the package's role.)

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/input-processor/src/correctionLayout.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+   * keyboard's layout bounding box mapped to a box from <0, 0> to <1, 1>.
+   */
+  readonly height: number;
+}
+
+export interface CorrectionLayout {
+  /**
+   * Defines the mappings of each key to be considered by a key-correction
+   * algorithm.  The key's bounding box should be defined relative to its
+   * containers bounding box, with both mapped to a coordinate system from
```
Comment (mcdurdin) [inline common/web/input-processor/src/correctionLayout.ts:43]:
```suggestion
   * container's bounding box, with both mapped to a coordinate system from
```

File: common/web/input-processor/src/corrections.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * @param touchCoords A proportional (x, y) coordinate of the touch within the keyboard's geometry.
+ *                           Should be within <0, 0> to <1, 1>.
+ * @param correctiveLayout  The corrective-layout mappings for keys under consideration
+ *                          by a correction algorithm, also within <0, 0> to <1, 1>.
+ * @returns A mapping of key IDs to the 'squared pseudo-distance' of the touchpoint to each key.
+ */
+export function keyTouchDistances(touchCoords: {x: number, y: number}, correctiveLayout: CorrectionLayout): Map<string, number> {
+  let keyDists: Map<string, number> = new Map<string, number>();
+
+  // This double-nested loop computes a pseudo-distance for the touch from each key.  Quite useful for
```
Comment (mcdurdin) [inline common/web/input-processor/src/corrections.ts:15]: I don't see the double-nesting?

What is the performance of this algorithm?

**Review by mcdurdin — APPROVED** [review]

File: common/web/input-processor/src/corrections.ts (hunk 2 above)
Comment (jahorton) [reply common/web/input-processor/src/corrections.ts:15]: Ah, right... the refactor split that up.  The double-nesting is from `correctionLayout.ts`, `buildCorrectiveLayout` - you can see a nested `.map()` call in that method.

```suggestion
  // This loop computes a pseudo-distance for the touch from each key. Quite useful for
```

Comment (jahorton) [reply common/web/input-processor/src/corrections.ts:15]: This specific function should be O(n) - as well as the `buildCorrectiveLayout` - where `n` is the total number of keys specified in the "corrective layout".

**Pull request merged**